File: src/common/wasm/zig/reduce_nansum.zig
//! WASM reduction nansum kernels for float types only.
//!
//! Reduction: result = sum(a[i] for i where !isnan(a[i]))
//! Int/uint types have no NaN — TS routes to regular sum.

/// Nansum f64 array, scalar (NaN check prevents SIMD vectorization).
export fn reduce_nansum_f64(a: [*]const f64, N: u32) f64 {
    var total: f64 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const v = a[i];
        if (v == v) total += v; // NaN != NaN
    }
    return total;
}

/// Nansum f32 array.
export fn reduce_nansum_f32(a: [*]const f32, N: u32) f32 {
    var total: f32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const v = a[i];
        if (v == v) total += v;
    }
    return total;
}

// --- Tests ---

test "reduce_nansum_f64 basic" {
    const testing = @import("std").testing;
    const nan = @as(f64, @bitCast(@as(u64, 0x7FF8000000000000)));
    const a = [_]f64{ 1.0, nan, 3.0, nan, 5.0 };
    try testing.expectApproxEqAbs(reduce_nansum_f64(&a, 5), 9.0, 1e-10);
}

test "reduce_nansum_f64 all NaN is zero" {
    const testing = @import("std").testing;
    const nan = @as(f64, @bitCast(@as(u64, 0x7FF8000000000000)));
    const a = [_]f64{ nan, nan, nan };
    try testing.expectApproxEqAbs(reduce_nansum_f64(&a, 3), 0.0, 1e-10);
}

test "reduce_nansum_f64 no NaN same as sum" {
    const testing = @import("std").testing;
    const a = [_]f64{ 1.0, 2.0, 3.0 };
    try testing.expectApproxEqAbs(reduce_nansum_f64(&a, 3), 6.0, 1e-10);
}

test "reduce_nansum_f64 NaN at start and end" {
    const testing = @import("std").testing;
    const nan = @as(f64, @bitCast(@as(u64, 0x7FF8000000000000)));
    const a = [_]f64{ nan, 5.0, 10.0, nan };
    try testing.expectApproxEqAbs(reduce_nansum_f64(&a, 4), 15.0, 1e-10);
}

test "reduce_nansum_f32 basic" {
    const testing = @import("std").testing;
    const nan = @as(f32, @bitCast(@as(u32, 0x7FC00000)));
    const a = [_]f32{ 2.0, nan, 3.0 };
    try testing.expectApproxEqAbs(reduce_nansum_f32(&a, 3), 5.0, 1e-5);
}

test "reduce_nansum_f32 all NaN is zero" {
    const testing = @import("std").testing;
    const nan = @as(f32, @bitCast(@as(u32, 0x7FC00000)));
    const a = [_]f32{ nan, nan };
    try testing.expectApproxEqAbs(reduce_nansum_f32(&a, 2), 0.0, 1e-5);
}

test "reduce_nansum_f64 empty" {
    const testing = @import("std").testing;
    const a = [_]f64{};
    try testing.expectApproxEqAbs(reduce_nansum_f64(&a, 0), 0.0, 1e-10);
}

File: src/common/wasm/reduce_nansum.ts
/**
 * WASM-accelerated reduction nansum.
 *
 * Reduction: result = nansum(a[0..N])
 * Returns null if WASM can't handle this case.
 * Only float64 and float32 — TS routes int/uint to regular sum.
 */

import { reduce_nansum_f64, reduce_nansum_f32 } from './bins/reduce_nansum.wasm';
import { resetScratchAllocator, resolveInputPtr } from './runtime';
import { ArrayStorage } from '../storage';
import type { DType, TypedArray } from '../dtype';
import { wasmConfig } from './config';

const BASE_THRESHOLD = 32;

type ReduceFn = (aPtr: number, N: number) => number | bigint;

const kernels: Partial<Record<DType, ReduceFn>> = {
  float64: reduce_nansum_f64,
  float32: reduce_nansum_f32,
  // float16 excluded: must accumulate in f16 precision to match NumPy overflow behavior
};

type AnyTypedArrayCtor = new (length: number) => TypedArray;
const ctorMap: Partial<Record<DType, AnyTypedArrayCtor>> = {
  float64: Float64Array,
  float32: Float32Array,
};

/**
 * WASM-accelerated reduction nansum (no axis, full array).
 * Returns null if WASM can't handle (complex types, non-contiguous, too small).
 */
export function wasmReduceNansum(a: ArrayStorage): number | null {
  if (!a.isCContiguous) return null;

  const size = a.size;
  if (size < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const dtype = a.dtype;
  const kernel = kernels[dtype];
  const Ctor = ctorMap[dtype];
  if (!kernel || !Ctor) return null;

  wasmConfig.wasmCallCount++;
  resetScratchAllocator();

  const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
  const aPtr = resolveInputPtr(a.data, a.isWasmBacked, a.wasmPtr, a.offset, size, bpe);

  return Number(kernel(aPtr, size));
}
